gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/PartitionedRegion.java

```diff
@@ -7172,9 +7176,11 @@ public boolean hasNext() {
             // KN: If DiskIterator has been initialized then we need to change
             // the current bucket id and the current bucket region accordingly
             // as otherwise there will be a mismatch
-            if (val != null && this.diskIteratorInitialized) {
+            if (val != null && !(val instanceof NonLocalRegionEntry) && this.diskIteratorInitialized) {
               setCurrRegionAndBucketId(val);
             }
+            //TODO: Suranjan How will it work for tx? There will be no BucketRegion for NonLocalRegionEntry
+            // Ideally for tx there shouldn't be a case of iterator fetching remote entry
```

Comment:
Right, for tx need to handle separately. It is actually a general problem also discussed during distributed joins (i.e. what to do for remote entries fetched from remote that may need to be read-locked for transactions). Ideal solution seems to be using snapshot isolation semantics for transactions that we need to do in any case. However, the need for read-locks in REPEATABLE_READ and SERIALIZABLE isolation levels is still there for proper conflict detection behaviour we provide.

One option is to pessimistically acquire read locks on all remotely fetched entries and then keep releasing locks at higher layers as those rows get disqualified from higher layer predicates (e.g. in SELECT * FROM T1 WHERE A=1 as the higher level filter A=1 gets applied and rows are rejected, they should release locks at that point).

An optimistic approach is to instead acquire remote read locks at higher layers only for qualified rows. This has trouble with complex workflows where entries may be getting inserted/deleted while the scan is going on. Snapshot isolation alleviates some of the issues but will result in larger number of conflicts.

Both options have been discussed for distributed joins. Both will be difficult to implement in the catalyst engine.

A third option yet to be fully thought through is using fine-grained snapshot locking. That is no read locks with a snapshot, then track only the entries that ultimately contributed to the result of that query (or multiple in the same transaction) -- again this will need changes to catalyst engine that can be somewhat complex. The RVVs for these entries can be checked at commit time for any changes at which point a conflict can be thrown. The downside of this approach is that readers can start getting conflicts which is not the case currently. However, it does have cleaner and better defined semantics.

For column tables things are simpler because there are no updates/deletes as of now, so there will be no read-write conflicts with snapshot isolation.